DFE-Digital/sts-plan-technology-for-your-school
# feat: Amend GTM - add HTML + JS to HTML and use IDs from config instead

## Overview

Addresses ticket 228927

Moves the GTM/Google Analytics HTML + JS to the web app, and use the appsettings.json for the IDs

## Changes

### Minor

#### Web app

- Move GTM + Analytics HTML + JS to web app
   - Create partial views for the relevant parts
   - Use IConfiguration for the ids
- Use the [options pattern](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-8.0) for the GTM configuration
- Create an `appsettings.ENVIRONMENT.json` per environment
- Tidy up the existing `appsettings.Development.json`; remove anything that is already identical to `appsettings.json`
- Tidied up the `Program.cs` some more
    - Create a new file/class for the non-development environment only services (i.e. KV and Data Protection _stuff_ currently)
    - Add almost all remaining services that were added into the chained call and/or moved to `ProgramExtensions.cs` if not done already
   
#### Terraform

- Add `ASPNETCORE_ENVIRONMENT` to the container app using a variable `container_environment` variable
- Add a `container_environment_variables` variable that allows settings whatever environment variables you want in the web app
   - Not actually used, but this was my initial approach, and I thought good for future proofing so I've left it in
- Moved the existing env variables we set for the container app to the locals as `container_environment_variables_default`, with the `container_environment`/`ASPNETCORE_ENVIRONMENT` variable added

### Non-functional changes 

- Add unit tests for cookies page, that ensures the user's chosen preferences are displayed in the form. I.e. if they choose "yes" then "yes" should be checked once the page loads, and if they choose "no" then "no" should be checked once the page loads
    - Also test here if the GTM code is/isn't showing on the page as expected

## How to review the PR

1. Run the web app
2. Enable cookies
3. Assert that the `meta` element, the `script` element, and the `noscript` element exist in the source
4. Assert that the ids in the configuration match the ones in the HTML. These are currently set locally as dummy values in the `appsettings.json`, and are overriden by the KV.
6. Disable cookies
7. Verify that the same elements no longer exist

## Checklist

Delete any rows that do not apply to the PR.

- [x] Terraform has been updated
